crates/sui-indexer-alt/src/ingestion/regulator.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use mysten_metrics::spawn_monitored_task;
+use tokio::{sync::mpsc, task::JoinHandle};
+use tokio_util::sync::CancellationToken;
+use tracing::info;
+
+/// The regulator task is responsible for writing out checkpoint sequence numbers from the
+/// `checkpoints` iterator to `checkpoint_tx`, bounded by the high watermark dictated by
+/// subscribers.
+///
+/// Subscribers can share their high watermarks on `ingest_hi_rx`. The regulator remembers these,
+/// and stops serving checkpoints if they are over the minimum subscriber watermark plus the
+/// ingestion `buffer_size`.
+///
+/// This offers a form of back-pressure that is sensitive to ordering, which is useful for
+/// subscribers that need to commit information in order: Without it, those subscribers may need to
+/// buffer unboundedly many updates from checkpoints while they wait for the checkpoint that they
+/// need to commit.
+///
+/// Note that back-pressure is optional, and will only be applied if a subscriber provides a
+/// watermark, at which point it must keep updating the watermark to allow the ingestion service to
+/// continue making progress.
+///
+/// The task will shut down if the `cancel` token is signalled, or if the `checkpoints` iterator
+/// runs out.
+pub(super) fn regulator<I>(
+    checkpoints: I,
+    buffer_size: usize,
+    mut ingest_hi_rx: mpsc::UnboundedReceiver<(&'static str, u64)>,
+    checkpoint_tx: mpsc::Sender<u64>,
+    cancel: CancellationToken,
+) -> JoinHandle<()>
+where
+    I: IntoIterator<Item = u64> + Send + Sync + 'static,
+    I::IntoIter: Send + Sync + 'static,
+{
+    spawn_monitored_task!(async move {
+        let mut ingest_hi = None;
+        let mut subscribers_hi = HashMap::new();
+        let mut checkpoints = checkpoints.into_iter().peekable();
+
+        info!("Starting ingestion regulator");
+
+        loop {
+            let Some(cp) = checkpoints.peek() else {
```

Comment:
to make sure I follow, does this only execute when `u64::MAX` is done?

Reply:
It's possible to pass a "last_checkpoint` to the ingestion (e.g. for backfill), so it can fire then too.